src/core/geometry/qgslinestring.cpp

```diff
@@ -2387,3 +2387,166 @@ QgsLineString *QgsLineString::measuredLine( double start, double end ) const
+    return clone();
+
+  const double *xData = mX.constData();
+  const double *yData = mY.constData();
+  const double *mData = mM.constData();
+  const double *zData = is3D() ? mZ.constData() : nullptr;
+
+  QVector< double > xOut( totalPoints );
+  QVector< double > yOut( totalPoints );
+  QVector< double > mOut( totalPoints );
+  QVector< double > zOut( is3D() ? totalPoints : 0 );
+
+  double *xOutData = xOut.data();
+  double *yOutData = yOut.data();
+  double *mOutData = mOut.data();
+  double *zOutData = is3D() ? zOut.data() : nullptr;
+
+  int i = 0;
+  double currentSegmentLength = 0;
+  double lastValidM = std::numeric_limits< double >::quiet_NaN();
+  double prevX = *xData;
+  double prevY = *yData;
+  while ( i < totalPoints )
+  {
+    double thisX = *xData++;
+    double thisY = *yData++;
+    double thisZ = zData ? *zData++ : 0;
+    double thisM = *mData++;
+
+    currentSegmentLength = QgsGeometryUtilsBase::distance2D( prevX, prevY, thisX, thisY );
+
+    if ( !std::isnan( thisM ) )
+    {
+      *xOutData++ = thisX;
+      *yOutData++ = thisY;
+      *mOutData++ = thisM;
+      if ( zOutData )
+        *zOutData++ = thisZ;
+      lastValidM = thisM;
+    }
+    else if ( i == 0 )
+    {
+      // nan m value at start of line, read ahead to find first non-nan value and backfill
+      int j = 0;
+      double scanAheadM = thisM;
+      while ( i + j + 1 < totalPoints && std::isnan( scanAheadM ) )
+      {
+        scanAheadM = mData[ j ];
+        ++j;
+      }
+      if ( std::isnan( scanAheadM ) )
+      {
+        // no valid m values in line
+        return nullptr;
+      }
+      *xOutData++ = thisX;
+      *yOutData++ = thisY;
+      *mOutData++ = scanAheadM;
+      if ( zOutData )
+        *zOutData++ = thisZ;
+      for ( ; i < j; ++i )
+      {
+        thisX = *xData++;
+        thisY = *yData++;
+        *xOutData++ = thisX;
+        *yOutData++ = thisY;
+        *mOutData++ = scanAheadM;
+        mData++;
+        if ( zOutData )
+          *zOutData++ = *zData++;
+      }
+      lastValidM = scanAheadM;
+    }
+    else
+    {
+      // nan m value in middle of line, read ahead till next non-nan value and interpolate
+      int j = 0;
+      double scanAheadX = thisX;
+      double scanAheadY = thisY;
+      double distanceToNextValidM = currentSegmentLength;
+      std::vector< double > scanAheadSegmentLengths;
+      scanAheadSegmentLengths.emplace_back( currentSegmentLength );
+
+      double nextValidM = std::numeric_limits< double >::quiet_NaN();
+      while ( i + j < totalPoints - 1 )
+      {
+        double nextScanAheadX = xData[j];
+        double nextScanAheadY = yData[j];
+        double nextScanAheadM = mData[ j ];
+        const double scanAheadSegmentLength = QgsGeometryUtilsBase::distance2D( scanAheadX, scanAheadY, nextScanAheadX, nextScanAheadY );
+        scanAheadSegmentLengths.emplace_back( scanAheadSegmentLength );
+        distanceToNextValidM += scanAheadSegmentLength;
```

Comment:
If one x or y is NaN (rare case, yes), you'll propagate the NaN value

Reply:
> If one x or y is NaN (rare case, yes)

I'm not going to handle that case here -- it will complicate the logic considerably and none of our other geometry routines handle nan x/y values.

If this is a scenario we want to support then the user should first run some form of "remove nan x/y vertices" tool on their data and clean it up.

Reply:
I agree with you, my comment was to warn you that this function interpolates value to handle NaN, but you can have NaN via 2D Distance. Indeed, you don't have to handle the case here. However, since QGIS can support x/y NaN, I think that we should mention this case : a line in the documentation, for example.

Reply:
My 2 cents: I don't think this needs addressing here. Beyond the rarity of such scenario, we would need to address the issue consistently and systematically across already existing algorithms (such as the interpolate point by distance). If we want to address this one day, let's do it thoughtfully and in a separate PR.